src/Time/Actions/LimitTimeStepToExpirationTimes.hpp

```diff
@@ -0,0 +1,175 @@
+/// are both needed at each substep and updated only after taking a full step.
+/// This can result in quiescence, i.e., in waiting to complete the next
+/// substep until the functions of time are updated, but the functions of time
+/// will only be updated after the next step is taken. To avoid this situation,
+/// at the start of each full step, this action checks whether the current
+/// step size exceeds the time remaining before the first function of time
+/// expires. If it does, then the step size is adjusted to be earlier than
+/// the expiration time of the next function of time to expire. Note that
+/// if the time stepper does not use substeps, this action does nothing.
+///
+/// Uses:
+/// - DataBox:
+///   - Tags::Time
+///   - Tags::TimeStep
+///   - Tags::TimeStepId
+///   - Tags::TimeStepper<>
+///   - domain::Tags::FunctionsOfTime
+///
+/// DataBox changes:
+/// - Adds: nothing
+/// - Removes: nothing
+/// - Modifies:
+///   - Tags::TimeStep
+struct LimitTimeStepToExpirationTimes {
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTags>& box, tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {  // NOLINT const
+    // First, check whether the time stepper uses substeps
+    const TimeStepper& time_stepper = db::get<Tags::TimeStepper<>>(box);
+    if (time_stepper.number_of_substeps() > 0) {
+      // Next, check if the current time step is at the beginning of a full step
+      const auto& time_step_id = db::get<Tags::TimeStepId>(box);
+      if (time_step_id.substep() == 0) {
+        // Get the functions of time and, if not empty, loop over them,
+        // finding the minimum expiration time
+        const auto& functions_of_time =
+            get<domain::Tags::FunctionsOfTime>(cache);
+        if (not functions_of_time.empty()) {
+          double min_expiration_time{std::numeric_limits<double>::max()};
+          for (const auto& [name, f_of_t] : functions_of_time) {
+            if (f_of_t->time_bounds()[1] < min_expiration_time) {
+              min_expiration_time = f_of_t->time_bounds()[1];
+            }
+          }
+
+          ASSERT(db::get<Tags::Time>(box) <= min_expiration_time,
+                 "Time (" << std::setprecision(20) << db::get<Tags::Time>(box)
+                          << ") is greater than expiration time ("
+                          << min_expiration_time << ") by an amount of "
+                          << db::get<Tags::Time>(box) - min_expiration_time);
+          if (db::get<Tags::Time>(box) == min_expiration_time) {
+            return {Parallel::AlgorithmExecution::Continue, std::nullopt};
```

Comment:
We know this actually should be `Retry` with a FoT update callback.  At a minimum, put a comment here.

It would be good to make an attempt to figure out why that doesn't work, because if there is some fundamental issue we may need to find a different solution than this action when we get to faster control system scales.

Reply:
Agreed.